### PR TITLE
WIP: Initial UNET implementation

### PR DESCRIPTION
**Pull request by fzimmermann89**
Based on @mxlutz work of getting it into github and @SherineBrahma and @koflera 's work in progress.

This will still need some major cleanup, but as CK suggested, maybe we try to get some version of it into main till next week.

I rebased this to current main, i.e. only put the Unet commit into this branch to keep stuff a bit cleaner and easier to merge

### CODE REVIEW

File: src/inhomcorr/methods/unet.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+            x = self.upconvs[i](x)
+            x = torch.cat([x, enc_features], dim=1)
+            x = self.dec_blocks[i](x)
+        return x
+
+
+class UNet(nn.Module):
+    """UNet model."""
+
+    def __init__(self, hparams: HyperParameters):
```
Comment (schuenke) [inline src/inhomcorr/methods/unet.py:178]: should be UNetHyperparameters

File: src/inhomcorr/methods/unet.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    """A block of convolutional layers (1D, 2D or 3D)."""
+
+    def __init__(
+            self,
+            dim,
+            n_ch_in,
+            n_ch_out,
+            n_convs,
+            kernel_size=3,
+            bias=False):
```
Comment (schuenke) [inline src/inhomcorr/methods/unet.py:20]: bias should always be true imho

File: src/inhomcorr/methods/unet.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        super(ConvBlock, self).__init__()
```
Comment (schuenke) [inline src/inhomcorr/methods/unet.py:21]: super().__init__()

File: src/inhomcorr/methods/unet.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+        if dim == 1:
+            conv_op = nn.Conv1d
+        if dim == 2:
+            conv_op = nn.Conv2d
+        elif dim == 3:
+            conv_op = nn.Conv3d
+
+        padding = int(np.floor(kernel_size / 2))
```
Comment (schuenke) [inline src/inhomcorr/methods/unet.py:30]: padding='same'

File: src/inhomcorr/methods/unet.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+        self.conv_block = nn.Sequential(*conv_block_list)
+
+    def forward(self, x):
+        """Forward pass of the convolutional block.
+
+        Parameters
+        ----------
+        x
+            _description_
```
Comment (schuenke) [inline src/inhomcorr/methods/unet.py:53]: docstring

File: src/inhomcorr/methods/unet.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    def __init__(
+            self,
+            dim,
+            n_ch_in,
+            n_enc_stages,
+            n_convs_per_stage,
+            n_filters,
+            kernel_size=3,
+            bias=False):
+        super(Encoder, self).__init__()
```
Comment (schuenke) [inline src/inhomcorr/methods/unet.py:74]: see above.

File: src/inhomcorr/methods/unet.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+
+    def __init__(
+            self,
+            dim,
+            n_ch_in,
+            n_enc_stages,
+            n_convs_per_stage,
+            n_filters,
+            kernel_size=3,
+            bias=False):
```
Comment (schuenke) [inline src/inhomcorr/methods/unet.py:73]: see above

File: src/inhomcorr/methods/unet.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+            n_ch_list.append(int(n_filters) * 2**ne)
+
+        self.enc_blocks = nn.ModuleList([ConvBlock(dim,
+                                                   n_ch_list[i],
+                                                   n_ch_list[i + 1],
+                                                   n_convs_per_stage,
+                                                   kernel_size=kernel_size)
+                                         for i in range(len(n_ch_list) - 1)])
+
+        if dim == 1:
```
Comment (schuenke) [inline src/inhomcorr/methods/unet.py:87]: do we want to fix our self to 2d for sake of simplicity? and use a different class for 3d etc?

File: src/inhomcorr/methods/unet.py [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+        elif dim == 2:
+            conv_op = nn.Conv2d
+            interp_mode = 'bilinear'
+        elif dim == 3:
+            interp_mode = 'trilinear'
+            conv_op = nn.Conv3d
+
+        self.interp_mode = interp_mode
+
+        padding = int(np.floor(kernel_size / 2))
```
Comment (schuenke) [inline src/inhomcorr/methods/unet.py:146]: 'same'

File: src/inhomcorr/methods/unet.py [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+            conv_op = nn.Conv1d
+        elif hparams.dim == 2:
+            conv_op = nn.Conv2d
+        elif hparams.dim == 3:
+            conv_op = nn.Conv3d
+
+        self.c1x1 = conv_op(hparams.nFilters,
+                            hparams.nChOut,
+                            kernel_size=1, padding=0, bias=hparams.Bias)
+        self.res_connection = hparams.ResConnection
```
Comment (schuenke) [inline src/inhomcorr/methods/unet.py:207]: i would remove this residual connection from the parameters. this is not useful in the general case but can be used to the model using the unet if needed.

File: src/inhomcorr/methods/unet.py [hunk 11]
```diff
@@ -0,0 +1,228 @@
+"""Implementation of a UNet."""
+
+import numpy as np
+import torch
+import torch.nn as nn
+
+from inhomcorr.interfaces.bias_estimator_interface import HyperParameters
+
+
+class ConvBlock(nn.Module):
```
Comment (schuenke) [inline src/inhomcorr/methods/unet.py:10]: maybe replace by a resnet block?

File: src/inhomcorr/methods/unet_estimator.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+class UNetHyperParameters(HyperParameters):
+    """Hyperparameters for the UNet Estimator."""
+
+    dim: torch.int = 2
+    nChIn: torch.int = 2
+    nChOut: torch.int = 2
+    KernelSize: torch.int = 3
+    nEncStages: torch.int = 3
+    nConvsPerStage: torch.int = 2
+    nFilters: torch.int = 16
```
Comment (schuenke) [inline src/inhomcorr/methods/unet_estimator.py:23]: make this configureable

**Review by schuenke — CHANGES_REQUESTED** [review]
some initial comments

File: src/inhomcorr/methods/unet.py [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+            x = torch.cat([x, enc_features], dim=1)
+            x = self.dec_blocks[i](x)
+        return x
+
+
+class UNet(nn.Module):
+    """UNet model."""
+
+    def __init__(self, hparams: UNetHyperParameters):
+        super(UNet, self).__init__()
```
Comment (fzimmermann89) [inline src/inhomcorr/methods/unet.py:215]: see above, use python 3` super().__init__()`

File: src/inhomcorr/methods/unet.py [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+                               (2**(hparams.nEncStages - 1)),
+                               hparams.nEncStages,
+                               hparams.nConvsPerStage,
+                               hparams.nFilters * (hparams.nEncStages * 2),
+                               hparams.ActivationFun,
+                               hparams.ConvBlockResConnection,
+                               kernel_size=hparams.KernelSize,
+                               bias=hparams.Bias)
+
+        if hparams.dim == 1:
```
Comment (fzimmermann89) [inline src/inhomcorr/methods/unet.py:237]: would still recommend removing the 1d/2d/3d switch everywhere...

File: src/inhomcorr/methods/unet.py [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+                            hparams.nChOut,
+                            kernel_size=1, padding=0, bias=hparams.Bias)
+
+    def forward(self, x):
+        """Forward pass of the UNet.
+
+        Parameters
+        ----------
+        x
+            _description_
```
Comment (fzimmermann89) [inline src/inhomcorr/methods/unet.py:254]: docstring

File: src/inhomcorr/methods/unet_estimator.py [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+        Parameters
+        ----------
+        image
+            Image with bias field
+
+        Returns
+        -------
+            Biasfield
+        """
+        biasfield = ImageData()
```
Comment (fzimmermann89) [inline src/inhomcorr/methods/unet_estimator.py:37]: switch to ImageData(image)

File: src/inhomcorr/methods/unet.py [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+
+    dim: torch.int = 2
+    nChIn: torch.int = 2
+    nChOut: torch.int = 2
+    ActivationFun: torch.nn = torch.nn.LeakyReLU()
+    KernelSize: torch.int = 3
+    nEncStages: torch.int = 3
+    nConvsPerStage: torch.int = 2
+    nFilters: torch.int = 16
+    ConvBlockResConnection: torch.bool = True
```
Comment (fzimmermann89) [inline src/inhomcorr/methods/unet.py:23]: types are wrong. keep int, bool etc! Those are supposed be scalars, not dtypes
